src/kernel/src/arch/aarch64/address.rs

```diff
@@ -1,119 +1,293 @@
-    pub fn new(_address: u64) -> Self {
+use super::memory::phys_to_virt;
+
+/// A representation of a canonical virtual address.
+#[derive(Clone, Copy, PartialEq, PartialOrd, Ord, Eq)]
+#[repr(transparent)]
+pub struct VirtAddr(u64);
+
+/// A representation of a valid physical address.
+#[derive(Clone, Copy, PartialEq, PartialOrd, Ord, Eq)]
+#[repr(transparent)]
+pub struct PhysAddr(u64);
+
+#[derive(Debug, Clone, Copy)]
+pub struct NonCanonical;
+
+impl VirtAddr {
+    /// The start of the kernel memory heap.
+    pub const HEAP_START: Self = Self(0xffffff0000000000);
+
+    pub const fn start_kernel_memory() -> Self {
         todo!()
     }
 
-    pub fn as_u64(self) -> u64 {
+    pub const fn start_user_memory() -> Self {
         todo!()
     }
 
-    pub fn align_up<U>(self, _alignment: U) -> Self
-    where
-        U: Into<u64>
-    {
+    pub const fn end_user_memory() -> Self {
         todo!()
     }
 
-    pub fn align_down<U>(self, _alignment: U) -> Self
-    where
-        U: Into<u64>
-    {
+    /// Construct a new virtual address from the provided addr value, only if the provided value is a valid, canonical
+    /// address. If not, returns Err.
+    pub const fn new(_addr: u64) -> Result<Self, NonCanonical> {
         todo!()
     }
-}
 
-#[derive(Clone, Copy, Debug)]
-pub struct VirtAddr;
+    /// Construct a new virtual address from a u64 without verifying that it is a valid virtual address.
+    ///
+    /// # Safety
+    /// The provided address must be canonical.
+    pub const unsafe fn new_unchecked(addr: u64) -> Self {
+        Self(addr)
+    }
 
-impl VirtAddr {
-    pub fn new(_address: u64) -> Self {
-        todo!()
+    pub fn as_mut_ptr<T>(&self) -> *mut T {
+        self.0 as *mut T
     }
 
-    pub fn as_u64(self) -> u64 {
-        todo!()
+    pub fn as_ptr<T>(&self) -> *const T {
+        self.0 as *const T
     }
 
-    pub fn from_ptr<T>(_ptr: *const T) -> Self {
-        todo!()
+    pub fn is_aligned_to(&self, alignment: usize) -> bool {
+        self.0 % alignment as u64 == 0
     }
 
-    pub fn as_ptr<T>(self) -> *const T {
+    pub fn is_kernel(&self) -> bool {
         todo!()
-    } 
+    }
 
-    pub fn as_mut_ptr<T>(self) -> *mut T {
-        todo!()
+    pub fn offset<U: Into<Offset>>(&self, offset: U) -> Result<Self, NonCanonical> {
+        let offset = offset.into();
+        match offset {
+            Offset::Usize(u) => Self::new(self.0.checked_add(u as u64).ok_or(NonCanonical)?),
+            Offset::Isize(u) => {
+                let abs = u.abs();
+                if u < 0 {
+                    Self::new(self.0.checked_sub(abs as u64).ok_or(NonCanonical)?)
+                } else {
+                    Self::new(self.0.checked_add(abs as u64).ok_or(NonCanonical)?)
+                }
+            }
+        }
     }
 
-    pub fn align_up<U>(self, _alignment: U) -> Self
-    where
-        U: Into<u64>
-    {
-        todo!()
+    pub fn raw(&self) -> u64 {
+        self.0
```

Comment:
Can this be named?  self.0 is just odd to me.

Reply:
The `.0` syntax is how you access values of a tuple. That is the naming convention in Rust.